Added ruff for linting and formatting

effaf opened the pull request:
Hopefully should resolve #3767 

Sorry for the mess created in the process. I am bit of a beginner here. Thank you for the patience

suhaibmujahid commented:
@effaf thank you for your interest in the project. 

The PR should resolve the issue, not another closed PR. I suggest reopening #4013 and addressing the comments there, specifically the following comments:
- https://github.com/mozilla/bugbug/pull/4013#issuecomment-1927285417 and 
- https://github.com/mozilla/bugbug/pull/4013#issuecomment-1931248402

If something is not working for you, you do not need to open a new PR. You can modify the same branch. Feel free to ask for help usg in the [bugbug](https://chat.mozilla.org/#/room/#bugbug:mozilla.org) Matrix room if you have any problems.

effaf commented:
I deleted that branch and repository. However, now I am sure I have fixed the linting error and all hooks successfully pass the test. Can I merge on this PR?

effaf commented:
It now successfully adds ruff to the project. Thank you for your patience. Please let me know if anything is missing.
resolves #4013

suhaibmujahid commented on .pre-commit-config.yaml:
```diff
@@ -1,25 +1,10 @@
-    rev: 23.11.0
-    hooks:
-      - id: black
-  - repo: https://github.com/pre-commit/mirrors-prettier
-    rev: v3.1.0
-    hooks:
-      - id: prettier
-        exclude: ^tests/fixtures/
```
Ruff does not replace Prettier, so we should keep this hook.

effaf replied:
Made the necessary changes. I see you have also made some changes. Let me know if I was wrong somewhere. Thank you for your guidance in this endeavor

suhaibmujahid reviewed (changes requested):
Thank you, @effaf!

Could you please replace the configuration files for `isort` and `flake8` with Ruff configurations similar to https://github.com/mozilla/bugbot/pull/2335?

Also, could you please modify the description of the PR so it resolves #3767 instead of #4013?

suhaibmujahid commented on .pre-commit-config.yaml:
```diff
@@ -1,25 +1,15 @@
 repos:
-  - repo: https://github.com/PyCQA/isort
-    rev: 5.12.0
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.2.1
```
Could you please move the `ruff` hook to be after the `prettier` hook? This will allow us to have a clearer diff.

effaf replied:
Done

suhaibmujahid commented on pyproject.toml:
```diff
@@ -0,0 +1,6 @@
+extend-exclude = ["data", "models", "db", "cache"]
+
+[tool.ruff.lint]
+select = ["E4", "E7", "E9", "F", "I", "T10", "CPY"]
```
Please add the `known-first-party` config for `isort`, which was already configured in the `.isort.cfg` file:
```suggestion
select = ["E4", "E7", "E9", "F", "I", "T10", "CPY"]

[tool.ruff.lint.isort]
known-first-party = ["bugbug_http"]
```

This should fix the unexpected changes in `http_service/tests/*`.

effaf replied:
Done

suhaibmujahid commented on pyproject.toml:
```diff
@@ -0,0 +1,6 @@
+# Ruff configurations
+[tool.ruff]
+extend-exclude = ["data", "models", "db", "cache"]
```
Based on the configurations in the `.flake8` file, we only need to exclude `data`. The `tools` directory is not used, and the rest (i.e., `.git` and `__pycache__`) should be already excluded with the default configs.
```suggestion
extend-exclude = ["data"]
```

effaf replied:
Done

suhaibmujahid commented on pyproject.toml:
```diff
@@ -0,0 +1,6 @@
+# Ruff configurations
```
This comment is not needed; the section header is self-explanatory:
```suggestion
```

effaf replied:
Done!

suhaibmujahid reviewed (changes requested):
Thank you! Please see my comments.

effaf commented:
added the resolutions as guided. 
please let me know if any changes are needed. Currently all the checks have passed. Thanks again @suhaibmujahid

suhaibmujahid commented on pyproject.toml:
```diff
@@ -0,0 +1,8 @@
+select = ["E4", "E7", "E9", "F", "I", "T10", "CPY"]
+
+[tool.ruff.lint.isort]
+known-first-party = ["bugbug_http"]
```
Could you please end the file with an empty line?
